src/plonk/circuit.rs

```diff
@@ -1072,11 +1156,280 @@ impl<F: Field> ConstraintSystem<F> {
+        // diagonal.
+        let mut exclusion_matrix = (0..self.num_selectors)
+            .map(|i| vec![false; i])
+            .collect::<Vec<_>>();
+
+        // Iterate over each of the selector columns
+        for (i, rows) in selectors.iter().enumerate() {
+            // Ignore exclusions with complex selectors
+            if complex_selectors[i] {
+                continue;
+            }
+
+            // Iterate over the rows, filtering for rows that are set
+            for set_index in rows.iter().enumerate().filter(|a| *a.1).map(|a| a.0) {
+                // Loop, looking only at the selectors previous to this one
+                for (j, other_selector) in selectors.iter().enumerate().take(i) {
+                    // Ignore exclusions with complex selectors
+                    if complex_selectors[j] {
+                        continue;
+                    }
+
+                    // Look at what selectors are active at the same row
+                    if other_selector[set_index] {
+                        // Mark them as incompatible
+                        exclusion_matrix[i][j] = true;
+                    }
+                }
+            }
+        }
+
+        // Keep track of the combinations; the ith element of this vector
+        // represents the ith combination and contains a vector of (simple)
+        // selectors that are included.
+        let mut combinations = vec![];
+
+        // We will not increase the degree of the constraint system, so we limit
+        // ourselves to the largest existing degree constraint.
+        let max_degree = self.degree();
+
+        // Simple selectors that we've added to combinations already.
+        let mut seen = vec![false; self.num_selectors];
+
+        let mut selector_map = vec![0; self.num_selectors];
+
+        for i in 0..self.num_selectors {
+            if complex_selectors[i] {
+                selector_map[i] = combinations.len();
+                // Complex selectors are on their own.
+                combinations.push(vec![i]);
+
+                // NB: We don't update `seen` here as it's a complex selector
+                // that will be ignored anyway.
+            } else {
+                // This is a simple selector.
+                if seen[i] {
+                    // This simple selector was added to a previous combination.
+                    continue;
+                }
+                selector_map[i] = combinations.len();
+                seen[i] = true; // We're definitely adding this to a combination.
+                let mut combination = vec![i];
+                assert!(degrees[i] <= max_degree); // Otherwise, self.degree() is wrong.
+
+                // Keep track of how large the substitution is.
+                let mut usage = 1;
+
+                // Keep track of how large of a gate any of the members of this
+                // combination can support.
+                let mut room = max_degree - degrees[i];
```

Comment:
The Python version has less code here because it isn't treating the first selector added to a combination as a special case (i.e. it's adding it in the loop over j). I don't see why it needs to be a special case?

Reply:
Oh, maybe it is slightly more efficient, because when you get to each combination you're not looping over selectors that have definitely already been seen. Yeah that's fine.